### ISSUE TITLE
onClickOutside doesn't respond to Composed component's updated props

### ISSUE THREAD

**Issue by 121watts** [Pomax/react-onclickoutside#159]
Hi all!  Thanks for the great package.  We use it all over the place in our app here @ influxdata.  

TLDR;

Should I be handling stateless components differently than in the example I provide?  
Or, could react-onclickoutside be updated to respond to composed component prop changes?

------
I have a stateless component that looks (kinda) like this:

```
const DatabaseRow = ({
  retentionPolicy,
}) => {
const {name, duration, replication, isEditing, isDefault, isNew} = retentionPolicy
  return (
    <div>
      `${name}${duration}${replication}${isEditing}${isDefault}${isNew}`
    </div>
  )
}

export default onClickOutside(DatabaseRow)
```

Which gets rendered by another component which looks a bit like this:

```
database.retentionPolicies.map(retentionPolicy => {
  return (
    <DatabaseRow
      key={rp.id}
      handleClickOutside={() => onStopEditRetentionPolicy({database, retentionPolicy})}
     />
  )
})
```

When I updated `retentionPolicy`, I'd expect `handleClickOutside` to create a new closure around the updated `retentionPolicy`.  So, when `handleClickOutside` fires `onStopEditRetentionPolicy` can operate on the most up to date `retentionPolicy`. 

Here's what I think is happening:

It looks like [this line](https://github.com/Pomax/react-onclickoutside/blob/master/index.js#L148) assigns the closure around `onStopeEditRetentionPolicy` (and its arguments) to the `clickOutsideHandler` variable on `componentDidMount`.  It looks like that variable is not updated / reassigned when the Composed component's props update.  I worked around this by copying all of the code in `componentDidMount` into `componentDidUpdate`.  Should I be handling stateless components differently?  Or, could react-onclickoutside be updated to respond to composed component prop changes?

**Comment by Pomax** [comment]
Hm, isn't the function signature for a stateless component `props => { ...; return ...; }`? It looks like your component actually follows a different signature, with an implicit object assignment that has the same property twice?

**Comment by 121watts** [comment]
@Pomax the syntax I used is a bit confusing, but is valid.  Apologies.  I updated it to a more standard syntax and this issue persists.  Thank you for the fast response!

**Comment by Pomax** [comment]
Hm, in the interest of a test, can you see what happens if you use
```
const DatabaseRow = (props) => {
  const {name, duration, replication, isEditing, isDefault, isNew} = props
  return (
    <div>
      `${name}${duration}${replication}${isEditing}${isDefault}${isNew}`
    </div>
  )
}

export default onClickOutside(Database
```
just to make sure it's not the object wrapping for the function argument that's causing problems?

**Comment by 121watts** [comment]
@Pomax Updated the code to use the syntax above.  Same issue.  : (

**Comment by Pomax** [comment]
I know it's asking a lot, but would you have time to turn that into a self-contained minimal demonstrator? If so, we can add it to the set of tests and then write a fix for it knowing that it won't regress in the future.

**Comment by Andarist** [comment]
Problem is that the actual handler is generated only in [`componentDidMount`](https://github.com/Pomax/react-onclickoutside/blob/f33aa72230404e9b7651399508deab1a7435a992/index.js#L172). So naturally if you change props later it wont get updated.

Fix should be quite straightforward, just wondering how to tackle this at code level. I think the best way is to compare with shallow equal all props on which the handler can depend in some lifecycle method and regenerate it if needed (+ ofc detach the old one and attach the new one). WDYT @Pomax ? I can provide a PR if we establish how this should be done.

**Comment by 121watts** [comment]
@Andarist that's exactly what I was seeing / thinking.  Thank you.

**Comment by Pomax** [comment]
That sounds reasonable, I see you have a PR open that fixes #161 already has changes to that effect, would that close this issue too if landed?

**Comment by Andarist** [comment]
Not rly. Separate PR needs to be provided. The other one only allows wrapped component to render null and setup and rip off document event listener accordingly

**Comment by Pomax** [comment]
gotcha - in that case yeah that approach sounds fine.

**Comment by Andarist** [comment]
@Pomax so I've wanted to fix this issue now - ofc `shallowEqual` approach would work, but then we end with quite some juggling with attaching/deattaching event handlers. 

So Ive started wondering why does the lib even generate the handler once - would be way easier if the handler just could access lib's `this` and always check the most recent values of `props` etc. Is there any reason why the handler was generated once? Ive tracked the git history and it was like that since initial commit.

Also I was wondering if we could simplify the whole "clicked outside" logic. Right now we traverse manually the DOM from the clicked element higher and higher in the hierarchy. Traversing could be done by the browsers themselves by using 
```js
var clickedOutside = !componentNode.contains(evt.target) && componentNode !== evt.target
```

**Comment by Pomax** [comment]
what is the support for `.contains` right now?

**Comment by Andarist** [comment]
According to - https://developer.mozilla.org/en-US/docs/Web/API/Node/contains it should be available pretty much everywhere, it doesnt even have its entry on caniuse.com.

> As for reversing the handler, can you expand on that?

I ment that we could just inline the handler when adding the listener, instead of generating it upfront and adding the generated version (which holds the values from the moment of creation, ignoring props etc updates)

**Comment by Pomax** [comment]
the up front generation is so that it doesn't constantly get reallocated every time something is unmounted and remounted. If there's an inlineish way to do that, I'll be happy to look at that.

**Comment by Andarist** [comment]
we could go with something like:
```js
var fn = evt => {
	if (this.props.disableOnClickOutside) {
		return;
	}
	// cache this in componentDidUpdate with this.componentNode = findDOMNode(this.getInstance());
	if (this.componentNode === null) {
		return
	}
	if (this.props.preventDefault) {
		evt.preventDefault();
	}
	if (this.props.stopPropagation) {
		evt.stopPropagation();
	}

	const current = evt.target;

	if (
		(this.props.excludeScrollbar && clickedScrollbar(evt)) ||
		  findHighest(current, this.componentNode, this.props.outsideClickIgnoreClass) !== document
		) {
      return;
    }
    // avoiding reallocations:
    // 1. the one coming from config may be cached for the life-time of the instance (like it is now)
    // 2. otherwise call instance.handleClickOutside(), no need for binding etc
    // as this.getInstance() is available in the closure
    // 3. fallbacking to instance.props.handleClickOutside()
    this.eventHandler(evt);

}
document.addEventListener(eventName, fn, handlerOptions);
```

do u see any downsides of this?

**Comment by Andarist** [comment]
@Pomax any thoughts on this?

**Comment by Pomax** [comment]
Hm, we can give this a shot and see what the tests bear out. Nothing immediately jumps out as possibly breaking here.
